Make partial= test helper take exactly two arguments

bshepherdson opened the pull request:
Fixes #21454.

`(= 2) ;=> true` and `partial=` followed that pattern. It leads to
confusing test assertions that silently succeed.

It's possible to extend this to support >= 2 arguments instead, but I'm
keeping it simple for now.

bshepherdson committed:
Make partial= test helper take exactly two arguments

Fixes #21454.

`(= 2) ;=> true` and `partial=` followed that pattern. It leads to
confusing test assertions that silently succeed.

It's possible to extend this to support >= 2 arguments instead, but I'm
keeping it simple for now.

dpsutton commented:
Two thoughts:
- you used a `:pre` check. Can you show what that error message looks like? 
- we don't ensure that there are only two args. Maybe capture `& too-many` and then equally assert `(empty? too-many)` so we don't think we are testing more values?

camsaul commented on test/metabase/test_runner/assert_exprs.clj:
```diff
@@ -118,10 +118,10 @@
-  `(do ~@(for [actual actuals]
-           `(t/do-report
-             (partial=-report ~message ~expected ~actual)))))
+  [message [_ expected actual]]
+  {:pre [(some? expected) (some? actual)]}
```
Previously this supported any number of args, e.g.

```clj
(is (partial= x y z))
```

which was basically the same as doing

```clj
(is (partial= x y))
(is (partial= x z))
```

After the changes here anything after the first two args (`z`) is just silently. That's not really much better than silently doing nothing with only one arg. I'm cool with only allowing exactly two args but we have to error if someone passes more than two, especially since it was previously totally fine.

This will also fail if you did

```clj
(is (partial= nil x))
```

since you're asserting `some?`. `partial= nil` is probably a dumb thing to be doing , but I don't think it's something that needs to be disallowed since the test will still _run_.

```suggestion
  [message [_ expected actual :as form]]
  {:pre [(= (count form) 3]}
```

I'd do it this way instead. count the `partial=` form itself. Note that `form` will be something like `(partial= x y)` so we want the count to be exactly `3`

camsaul reviewed (changes requested):
- This should fail if you pass in extra args instead of silently ignoring them since we used to allow extra args
- Checking count is better than checking `some?` IMO since `partial=` against `nil` might be dumb but probably doesn't need to be disallowed in the macroexpansion itself since it's not silently eating tests or anything like that

bshepherdson committed:
Rewire partial= to check the argument count properly

Also at execution time, not macro evaluation time, which seems to
produce nicer test failures.

bshepherdson commented:
The error (for the updated version) looks like
```
expected: (partial= {:a 1, :b []})
  actual: java.lang.AssertionError: Assert failed: partial= expects exactly 2 arguments
(clojure.core/= 2 1)
```

I've restructured this to be based on the arg count, and also to do the check inside the quote and not at macro expansion time, which produced cleaner error reports.

I tried to add tests for this but I can't find a permutation that actually captures the inner error. Any pointers for that?

camsaul commented on test/metabase/test_runner/assert_exprs.clj:
```diff
@@ -118,10 +118,11 @@
-  `(do ~@(for [actual actuals]
-           `(t/do-report
-             (partial=-report ~message ~expected ~actual)))))
+  [message [_ expected actual :as form]]
+  `(do
+     (assert (= 2 ~(count (rest form))) "partial= expects exactly 2 arguments")
+     (t/do-report
+       (partial=-report ~message ~expected ~actual))))
```
IMO I would do the assertion during the macroexpansion itself rather than in the expanded form. Otherwise we wouldn't see the error until we actually tried to run the test. IMO it's better to catch these types of syntax errors as we read them in rather than waiting until the test is ran

```suggestion
  [message [_ expected actual :as form]]
  (assert (= (count (rest form)) 2) "partial= expects exactly 2 arguments")
  `(t/do-report
    (partial=-report ~message ~expected ~actual)))
```

Just my 2 cents tho. I can see an argument for doing it the other way too.

Either way `(= 2 (count ...))` is a bit of a [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions) and I'd flip those around for readability

camsaul reviewed (approved):
I have some weakly-held opinions about doing the assertion during macroexpand time rather than when we actually run the test so fix that if you agree. If not then I'm ok with how it is.

Either way no need to request another review, just commit and hit merge once it's green

bshepherdson committed:
Move the assert back to macro time

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

dpsutton commented:
You are throwing an assertion on `(= 2 1)` which doesn't have much context and getting that in the error message which is less than ideal. Perhaps something like this would be a bit better?

```clojure
(defmethod t/assert-expr 'partial=
  [message [_ expected actual & extra-clauses :as form]]
  `(do
     (when (not= ~(count form) 3)
       (throw (ex-info "partial= expects exactly 2 arguments"
                       {:expected-form '~expected, :actual-form '~actual, :extra-clauses '~extra-clauses})))
     (t/do-report
      (partial=-report ~message ~expected ~actual))))
```

which yields

```clojure
assert-exprs=> (t/deftest foo
                 (t/is (partial= 3 (+ 1 1) (+ 2 2) (/ 4 1)))
                 (t/is (partial= 3))
                 (t/is (partial= 3 (+ 1 1)))
                 (t/is (partial= 3 (+ 1 2))))
#'metabase.test-runner.assert-exprs/foo
assert-exprs=> (foo)

ERROR in (foo) (NO_SOURCE_FILE:697)
expected: (partial= 3 (+ 1 1) (+ 2 2) (/ 4 1))
  actual: clojure.lang.ExceptionInfo: partial= expects exactly 2 arguments
{:extra-clauses ((+ 2 2) (/ 4 1)), :actual-form (+ 1 1), :expected-form 3}
 at [stacktrace ommitted]

ERROR in (foo) (NO_SOURCE_FILE:698)
expected: (partial= 3)
  actual: clojure.lang.ExceptionInfo: partial= expects exactly 2 arguments
{:extra-clauses nil, :actual-form nil, :expected-form 3}
 at [stacktrace ommitted]

FAIL in (foo) (NO_SOURCE_FILE:699)
expected: 3
  actual: 2
    diff: - 3
          + 2
nil
assert-exprs=> 
```

As for tests, I wouldn't worry about it. It's very straightforward.

dpsutton commented:
ah, ignore me. Moving it to compile time is a good call and then we don't have easy access to any `ex-data`. Sorry to backseat driver and nitpick. I'm glad you fixed this up.

camsaul commented:
@dpsutton The assertion does have the message

```
partial= expects exactly 2 arguments
```

which should be more than enough to figure out what when wrong especially since it will result in a CompilerException with the line where the offending form is.

With

```clj
(defmethod assert-expr 'my=
  [message [_ expected actual :as form]]
  (assert (= (count form) 3) "WRONG ARG COUNT BUDDY")
  `(do :ok))

(deftest x-test
  (is (my= 1 2))
  (is (my=))) ; line 299
```

I get the error 

```
2. Unhandled clojure.lang.Compiler$CompilerException
   Error compiling test/metabase/util/ssh_test.clj at (299:3)
   #:clojure.error{:phase :macroexpansion,
                   :line 299,
                   :column 3,
                   :source "/home/cam/metabase/test/metabase/util/ssh_test.clj",
                   :symbol clojure.test/try-expr}
             Compiler.java: 7023  clojure.lang.Compiler/macroexpand1
             Compiler.java: 7097  clojure.lang.Compiler/analyzeSeq
             Compiler.java: 6793  clojure.lang.Compiler/analyze
             Compiler.java: 7099  clojure.lang.Compiler/analyzeSeq
             Compiler.java: 6793  clojure.lang.Compiler/analyze
             Compiler.java: 7099  clojure.lang.Compiler/analyzeSeq
             Compiler.java: 6793  clojure.lang.Compiler/analyze
             Compiler.java: 6749  clojure.lang.Compiler/analyze
             Compiler.java: 6124  clojure.lang.Compiler$BodyExpr$Parser/parse
             Compiler.java: 5471  clojure.lang.Compiler$FnMethod/parse
             Compiler.java: 4033  clojure.lang.Compiler$FnExpr/parse
             Compiler.java: 7109  clojure.lang.Compiler/analyzeSeq
             Compiler.java: 6793  clojure.lang.Compiler/analyze
             Compiler.java: 7099  clojure.lang.Compiler/analyzeSeq
             Compiler.java: 6793  clojure.lang.Compiler/analyze
             Compiler.java: 6749  clojure.lang.Compiler/analyze
             Compiler.java: 3108  clojure.lang.Compiler$MapExpr/parse
             Compiler.java: 6801  clojure.lang.Compiler/analyze
             Compiler.java: 6749  clojure.lang.Compiler/analyze
             Compiler.java:  594  clojure.lang.Compiler$DefExpr$Parser/parse
             Compiler.java: 7111  clojure.lang.Compiler/analyzeSeq
             Compiler.java: 6793  clojure.lang.Compiler/analyze
             Compiler.java: 6749  clojure.lang.Compiler/analyze
             Compiler.java: 7185  clojure.lang.Compiler/eval
             Compiler.java: 7640  clojure.lang.Compiler/load
                      REPL:    1  metabase.util.ssh-test/eval134004
                      REPL:    1  metabase.util.ssh-test/eval134004
             Compiler.java: 7181  clojure.lang.Compiler/eval
             Compiler.java: 7136  clojure.lang.Compiler/eval
                  core.clj: 3202  clojure.core/eval
                  core.clj: 3198  clojure.core/eval
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 1977  clojure.core/with-bindings*
                  core.clj: 1977  clojure.core/with-bindings*
               RestFn.java:  425  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  152  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  218  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  217  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  833  java.lang.Thread/run

1. Caused by java.lang.AssertionError
   Assert failed: WRONG ARG COUNT BUDDY (= (count form) 3)

              ssh_test.clj:  294  metabase.util.ssh-test/eval134155/fn
```

which makes it perfectly easy to fix.

The benefit of doing the assert during macroexpansion is I don't have to wait to run `x-test` to actually see the error... I'll see it immediately when it reads that form in. It's basically a static assertion this way instead of a dynamic one
